File: src/main/java/com/github/happylynx/prick/cli/Command.java
package com.github.happylynx.prick.cli;

import static com.github.happylynx.prick.cli.CommandKt.syncCommand;

// TODO https://picocli.info/
class Command {

    private static final int GENERAL_ERROR_CODE = 1;

    void run(String[] args) {
        if (args.length == 0) {
            helpCommand();
            return;
        }
        final String firstArgument = args[0];
        if (firstArgument == null || "help".equals(firstArgument) || "--help".equals(firstArgument)) {
            helpCommand();
            return;
        }
        if ("--version".equals(firstArgument)) {
            versionOptionCommand(args);
            return;
        }
        if ("sync".equals(firstArgument)) {
            syncCommand(args);
            return;
        }
        if ("init".equals(firstArgument)) {
            CommandKt.initCommand(args);
            return;
        }
        unknownCommand(args);
    }

    private void versionOptionCommand(String[] args) {
        System.out.println("0.1");
    }

    private void unknownCommand(String[] args) {
        System.out.println("Unknown command: " + String.join(" ", args));
        helpCommand();
        CliUtils.INSTANCE.die();
    }

    private void helpCommand() {
        System.out.println("Usage: price [--help] [--version]");
        System.out.println("             commands [args...]");
        System.out.println("Commands:");
        System.out.println();
        System.out.println("    help    Print command usage help");
        System.out.println("    init    Initialize directory to be a synchronization root");
        System.out.println("    sync    Synchronizes selected directories");
    }

}

File: src/main/kotlin/com/github/happylynx/prick/cli/Command.kt
package com.github.happylynx.prick.cli

import com.github.happylynx.prick.lib.LibUtils
import com.github.happylynx.prick.lib.commands.InitCommand
import com.github.happylynx.prick.lib.commands.PrickContext
import com.github.happylynx.prick.lib.commands.SnapshotCommand
import com.github.happylynx.prick.lib.commands.SyncComamnd
import java.lang.IllegalStateException
import java.nio.file.Path

fun syncCommand(args: Array<String>) {
    // TODO add --subtree option
    if (args.size < 3) {
        Help.sync()
        CliUtils.die()
    }
    val dirs = args.drop(1)
            .map { arg ->
                val path = Path.of(arg)
                if (!LibUtils.isPrickRoot(path)) {
                    CliUtils.die("Path '$path' doesn't reference a prick directory.")
                }
                path.toAbsolutePath()
            }
    SyncComamnd(dirs, Path.of(".")).run()
}

fun initCommand(args: Array<String>) {
    val dir: Path = when (args.size) {
        1 -> Path.of(".")
        2 -> Path.of(args[1])
        else -> {
            Help.init(System.err)
            CliUtils.die()
            throw IllegalStateException()
        }
    }
    InitCommand(dir).run()

}

private fun throwEverytime() {
    throw RuntimeException()
}

fun snapshotCommand(args: Array<String>) {
    // TODO add --prick-root option
    if (args.size > 1) {
        CliUtils.die("Too many arguments of command snapshot")
    }
    val subtree = Path.of(".")
    val ctx = PrickContext.fromWorkingPath(subtree)
    SnapshotCommand(ctx, ctx.rootDir.relativize(subtree)).run()
}
